flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/fusion/spec/HashJoinFusionCodegenSpec.scala

```diff
@@ -0,0 +1,542 @@
+    }
+
+    getOperatorCtx.startNewLocalVariableStatement(buildRow)
+    val buildVars = genInputVars(buildRow, buildType)
+    val checkCondition = if (joinSpec.getNonEquiCondition.isPresent) {
+      // bind the build row name again
+      val expr = getExprCodeGenerator.generateExpression(joinSpec.getNonEquiCondition.get)
+      val skipRow = s"${expr.nullTerm} || !${expr.resultTerm}"
+      s"""
+         |// generate join condition
+         |${expr.getCode}
+         |if (!($skipRow))
+       """.stripMargin
+    } else {
+      ""
+    }
+    val buildLocalVars =
+      if (hashJoinType.leftSemiOrAnti() && !joinSpec.getNonEquiCondition.isPresent) {
+        ""
+      } else {
+        getOperatorCtx.reuseLocalVariableCode(buildRow)
+      }
+
+    (buildRow, checkCondition, buildLocalVars, buildVars)
+  }
+
+  /** Generates build side expr for outer join. */
+  protected def genProbeOuterBuildVars(
+      buildRow: String,
+      buildVars: Seq[GeneratedExpression]): Seq[GeneratedExpression] = {
+    buildVars.zipWithIndex.map {
+      case (expr, i) =>
+        val fieldType = buildType.getTypeAt(i)
+        val resultTypeTerm = primitiveTypeTermForType(fieldType)
+        val defaultValue = primitiveDefaultValue(fieldType)
+        val Seq(fieldTerm, nullTerm) =
+          getOperatorCtx.addReusableLocalVariables((resultTypeTerm, "field"), ("boolean", "isNull"))
+        val code = s"""
+                      |$nullTerm = true;
+                      |$fieldTerm = $defaultValue;
+                      |if ($buildRow != null) {
+                      |  ${expr.getCode}
+                      |  $nullTerm = ${expr.nullTerm};
+                      |  $fieldTerm = ${expr.resultTerm};
+                      |}
+          """.stripMargin
+        GeneratedExpression(fieldTerm, nullTerm, code, fieldType)
+    }
+  }
+
+  /** Generates the input row variables expr. */
+  def genInputVars(inputRowTerm: String, inputType: RowType): Seq[GeneratedExpression] = {
+    val indices = fieldIndices(inputType)
+    val buildExprs = indices
+      .map(
+        index => GenerateUtils.generateFieldAccess(getOperatorCtx, inputType, inputRowTerm, index))
+      .toSeq
+    indices.foreach(
+      index =>
+        getOperatorCtx
+          .addReusableInputUnboxingExprs(inputRowTerm, index, buildExprs(index)))
+
+    buildExprs
+  }
+
+  override def usedInputVars(inputId: Int): util.Set[Integer] = {
+    if (inputId == buildInputId) {
+      val set: util.Set[Integer] = new util.HashSet[Integer]()
+      buildKeys.toStream.map(key => set.add(key))
+      set
+    } else {
+      super.usedInputVars(inputId)
+    }
+  }
+
+  override def getInputRowDataClass(inputId: Int): Class[_ <: RowData] = {
+    if (inputId == buildInputId) {
+      // To build side, we wrap it BinaryRowData
+      classOf[BinaryRowData]
+    } else {
+      super.getInputRowDataClass(inputId)
+    }
+  }
+
+  private def codegenHashTable(spillEnabled: Boolean): Unit = {
```

Comment:
Do we have a plan to reuse most of the code with `LongHashJoinGenerator`?

Reply:
Yes. Since there is a relatively large difference between OFCG and single operator codegen in terms of doing code generation, the first step is not very good for us to do code reuse. Also due to 1.18 codefreeze, there is not much time to consider code reuse, but I believe this thing in subsequent versions we are going to do gradually, after all, the cost of maintaining two sets of code is huge.

Reply:
Thank you, could you open an issue for this under the umbrella issue please?

Reply:
create it https://issues.apache.org/jira/browse/FLINK-32510